### PR TITLE
PP-5354 Insert GovUkPayEvents

### PR DESCRIPTION
**Pull request by stephencdaly**
Insert GovUkPayEvents at the appropriate times.
Move event service classes to the correct package.

### CODE REVIEW

File: src/test/java/uk/gov/pay/directdebit/payments/services/GovUkPayEventServiceTest.java [hunk 1]
```diff
@@ -5,152 +5,157 @@
...
+        when(govUkPayEventStateGraph.isValidStartValue(eventType)).thenReturn(false);
 
         expectedException.expect(InvalidGovUkPayEventInsertionException.class);
         expectedException.expectMessage("GOV.UK Pay event MANDATE_SUBMITTED is invalid when there are no previous events");
 
-        govUkPayEventService.storeEvent(newEvent);
-    }
-
-    @Test
-    public void insertMandateEvent_shouldThrowForMissingMandateId() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/directdebit/payments/services/GovUkPayEventServiceTest.java]: Why did this go?

File: src/main/java/uk/gov/pay/directdebit/events/services/GovUkPayEventService.java [hunk 2]
```diff
@@ -26,32 +25,24 @@ public GovUkPayEventService(GovUkPayEventDao govUkPayEventDao,
...
-        govUkPayEventDao.findLatestEventForMandate(mandateId)
+    public void storeEventForMandate(Mandate mandate, GovUkPayEvent.GovUkPayEventType eventType) {
+        var event = new GovUkPayEvent(mandate, eventType);
+        govUkPayEventDao.findLatestEventForMandate(mandate.getId())
                 .ifPresentOrElse(latestEvent -> validateEventTransition(event, latestEvent),
                         () -> validateInitialEvent(event));
         
         govUkPayEventDao.insert(event);
-        LOGGER.info("Inserted GOV.UK Pay event for mandate {}", mandateId);
+        LOGGER.info("Inserted GOV.UK Pay event for mandate {}", mandate.getId());
```
[the remark this answers is not in the document]
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/events/services/GovUkPayEventService.java]: I changed the constructors for GovUkPayEvent, so it's no longer possible to have a null mandate id

File: src/test/java/uk/gov/pay/directdebit/payments/services/GovUkPayEventServiceTest.java (hunk 1 above)
Comment (stephencdaly) [reply src/test/java/uk/gov/pay/directdebit/payments/services/GovUkPayEventServiceTest.java]: I changed the constructors for GovUkPayEvent, so it's no longer possible to have a null mandate id

File: src/main/java/uk/gov/pay/directdebit/events/services/GovUkPayEventService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/events/services/GovUkPayEventService.java]: done

File: src/main/java/uk/gov/pay/directdebit/events/model/GovUkPayEvent.java [hunk 3]
```diff
@@ -1,30 +1,46 @@
...
-        this.id = id;
-        this.mandateId = mandateId;
-        this.paymentId = paymentId;
-        this.eventDate = Objects.requireNonNull(eventDate);
-        this.resourceType = Objects.requireNonNull(resourceType);
-        this.eventType = Objects.requireNonNull(eventType);
+    public GovUkPayEvent(Mandate mandate, GovUkPayEventType eventType) {
+        this.mandateId = mandate.getId();
+        this.resourceType = ResourceType.MANDATE;
+        this.eventDate = ZonedDateTime.now();
```
[the remark this answers is not in the document]
Comment (stephencdaly) [reply src/main/java/uk/gov/pay/directdebit/events/model/GovUkPayEvent.java]: done

File: src/main/java/uk/gov/pay/directdebit/events/services/GovUkPayEventService.java [hunk 4]
```diff
@@ -26,32 +25,24 @@ public GovUkPayEventService(GovUkPayEventDao govUkPayEventDao,
...
-            storeEventForPayment(event);
-        }
-    }
-
-    private void storeEventForMandate(GovUkPayEvent event) {
-        Long mandateId = event.getMandateId().orElseThrow(() -> new GovUkPayEventHasNoMandateIdException(event.getId()));
-        govUkPayEventDao.findLatestEventForMandate(mandateId)
+    public void storeEventForMandate(Mandate mandate, GovUkPayEvent.GovUkPayEventType eventType) {
+        var event = new GovUkPayEvent(mandate, eventType);
+        govUkPayEventDao.findLatestEventForMandate(mandate.getId())
```
[the remark this answers is not in the document]
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/directdebit/events/services/GovUkPayEventService.java]: As discussed, this is close to the DAO so it’s actually OK.

**Pull request merged**